src/test/test-utils.ts

```diff
@@ -0,0 +1,214 @@
+/// <reference path="../../typings/node/node.d.ts"/>
+/// <reference path="../../typings/tmp/tmp.d.ts"/>
+/// <reference path="../../typings/mocha/mocha.d.ts"/>
+/// <reference path="../../typings/chai/chai.d.ts"/>
+
+import * as net from 'net';
+import * as os from 'os';
+import * as fs from 'fs';
+
+import * as tmp from 'tmp';
+import * as mocha from 'mocha';
+import * as chai from 'chai';
+
+import * as vscode from 'vscode-languageserver';
+
+import Connection from '../connection';
+import {FileInfo} from '../fs';
+import * as rt from '../request-type';
+import * as utils from './test-utils';
+
+class Channel {
+    server: net.Server;
+    serverIn: net.Socket;
+    serverOut: net.Socket;
+    serverFile: string;
+    serverConnection: Connection;
+
+    client: net.Server;
+    clientIn: net.Socket;
+    clientOut: net.Socket;
+    clientFile: string;
+    clientConnection: Connection;
+}
+
+let channel: Channel;
+
+export function setUp(memfs: any, done: (err?: Error) => void) {
+
+    channel = new Channel();
+
+    const input = tmp.fileSync();
+    const output = tmp.fileSync();
+
+    fs.unlinkSync(input.name);
+    fs.unlinkSync(output.name);
+
+    channel.serverFile = input.name;
+    channel.clientFile = output.name;
+
+    let counter = 2;
+
+    function maybeDone() {
+        counter--;
+        if (counter === 0) {
+            channel.serverConnection.start();
+            channel.clientConnection.start();
+
+            const params: vscode.InitializeParams = {
+                processId: null,
+                rootPath: 'file:///',
+                capabilities: null
+            }
+
+            channel.clientConnection.sendRequest(rt.InitializeRequest.type, params).then(function () {
```

Comment:
This applies to the whole codebase, but why not use ES6 `() => ...` syntax? `function() { ... }` is much more verbose and has worse handling for `this` in most cases.